### PR TITLE
feat(shares): initial api implementation

### PR DESCRIPTION
**Pull request by kschelonka**
Initial shares service -- implementation following the agreed contract.

Query to retrieve data from a share slug and mutation to create a new link (authenticated users only).

Related updates: 
- DynamoDB TTL configuration in terraform modules
- Setting up the dynamodb in our localstack config

### CODE REVIEW

File: servers/shares-api/schema.graphql [hunk 1]
```diff
@@ -1,3 +1,72 @@
...
+
+scalar Url
+scalar Markdown
+scalar ISOString
+
 type PocketShare {
-  id: ID!
-}
\ No newline at end of file
+  slug: ID!
```
Comment (kschelonka) [inline servers/shares-api/schema.graphql:18]: This also contains `preview: ItemSummary` (federated via parser-graphql-wrapper)

Comment (bassrock) [reply servers/shares-api/schema.graphql:18]: I'm gonna rename it to preview in my pr. This name makes more sense.

File: servers/parser-graphql-wrapper/schema.graphql [hunk 2]
```diff
@@ -480,4 +480,9 @@ type ReaderInterstitial {
 
 type ItemNotFound {
   message: String 
-}
\ No newline at end of file
+}
+
+type PocketShare @key(fields: "targetUrl") {
+  targetUrl: Url!
+  preview: ItemSummary
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/schema.graphql:487]: alterantive: name `itemCard` to be more similar to ReaderView

File: servers/shares-api/schema.graphql [hunk 3]
```diff
@@ -1,3 +1,77 @@
...
\ No newline at end of file
+extend schema
+  @link(url: "https://specs.apollo.dev/link/v1.0")
+  @link(
+    url: "https://specs.apollo.dev/federation/v2.3"
+    import: ["@key", "@composeDirective", "@tag"]
+  )
+  # The link directive is required to federate @constraint
+  # It doesn't actually have to be a real spec, but it would be good
+  # to write one and replace this.
```
Comment (bassrock) [inline servers/shares-api/schema.graphql:9]: Surprised the library we use doesnt have one up

File: servers/parser-graphql-wrapper/schema.graphql (hunk 2 above)
Comment (bassrock) [inline servers/parser-graphql-wrapper/schema.graphql:487]: This federation needs to be enabled on the parser side yea?

**Review by bassrock — APPROVED** [review]
LGTM

File: servers/parser-graphql-wrapper/schema.graphql (hunk 2 above)
Comment (kschelonka) [reply servers/parser-graphql-wrapper/schema.graphql:487]: Yep, it's on the parser subgraph

File: servers/shares-api/schema.graphql (hunk 3 above)
Comment (kschelonka) [reply servers/shares-api/schema.graphql:9]: eh the readme of the repo probably counts 😅

File: servers/shares-api/schema.graphql [hunk 4]
```diff
@@ -1,3 +1,77 @@
...
+type ShareContext {
+  """
+  A user-provided comment/note on the shared content.
+  """
+  note: Markdown
+  """User-provided highlights of the content"""
+  highlights: [ShareHighlight!]
+}
+
+type ShareHighlight {
```
Comment (Gio2018) [inline servers/shares-api/schema.graphql:34]: wondering: how is this expected to render on the client?

Comment (kschelonka) [reply servers/shares-api/schema.graphql:34]: I think like a list of quotes (which is why I didn't include the article reconstruction context)

See upnext for prior art. I didn't add notes on the ShareHighlgiht yet to reduce complexity but we could

image

**Review by collectedmind — APPROVED** [review]

File: servers/shares-api/schema.graphql (hunk 4 above)
Comment (Gio2018) [reply servers/shares-api/schema.graphql:34]: thanks for clarifying @kschelonka !

**Review by Gio2018 — APPROVED** [review]

**Review by marcin-kozinski — APPROVED** [review]
Looks sensible to me 👍

File: servers/shares-api/schema.graphql (hunk 3 above)
Comment (kschelonka) [reply servers/shares-api/schema.graphql]: I tried this but the link directive requires a specific URL which does not pass with a semver like `vX.X.X` (requires `vX.X`)

File: servers/shares-api/schema.graphql [hunk 5]
```diff
@@ -1,3 +1,88 @@
...
+  """
+  Highlighted text on a piece of shared content.
+  This is a permissive constraint but there needs
+  to be _a_ constraint.
+  This input is not required, but if present 'quotes'
+  is required as it is the only field.
+  Limited to 300 characters per quote (longer quotes
+  will not be rejected, but will be truncated).
+  """
+  quotes: [Max300CharString!]! @constraint(maxItems: 20)
```
Comment (bassrock) [inline servers/shares-api/schema.graphql:53]: I like the custom scalar here

File: servers/shares-api/src/apollo/context.ts [hunk 6]
```diff
@@ -16,8 +22,27 @@ export async function getContext({
...
+  PocketShareModel: PocketShareModel;
+  constructor(options: { request: Request }) {
+    const userId = options.request.headers.userid;
+    const client = dynamoClient();
+    // We have an authenticated user?
+    if (
+      userId &&
+      typeof userId === 'string' &&
+      userId.length &&
+      userId !== 'anonymous'
```
Comment (bassrock) [inline servers/shares-api/src/apollo/context.ts:39]: lets ticket making this logic and logic from the context headers into a package.

File: servers/shares-api/src/config/index.ts [hunk 7]
```diff
@@ -24,13 +24,16 @@ export default {
     maxBackoff: 3000, // in ms, max amount of backoff time allowed for multiple requests
   },
   dynamoDb: {
-    notesTable: {
+    sharesTable: {
       name: process.env.SHARES_TABLE || 'SHARES-local-shares',
-      key: process.env.SHARES_TABLE_KEY || 'shareId',
+      ttl: 365 * 24 * 60 * 60 * 1000, // ~1 year in ms
```
Comment (bassrock) [inline servers/shares-api/src/config/index.ts:29]: 👍

**Review by bassrock — APPROVED** [review]
straight forward, just a few followups

File: servers/shares-api/src/apollo/context.ts (hunk 6 above)
Comment (kschelonka) [reply servers/shares-api/src/apollo/context.ts:39]: https://mozilla-hub.atlassian.net/browse/POCKET-9973

**Pull request merged**